### PR TITLE
Feature/ButtonPress

### PR DESCRIPTION
**Pull request by LitvinenkoIra**
Implemented ButtonPress request base functionality

### CODE REVIEW

File: src/components/can_cooperation/src/commands/button_press_request.cc [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                                  (button_name == enums_value::kDefrostRear) ||
+                                  (button_name == enums_value::kUpperVent) ||
+                                  (button_name == enums_value::kLowerVent);
+  }
+
+  if (module_type_and_button_name) {
+    SendRequest(
+        functional_modules::hmi_api::button_press, request_params, true);
+  } else {
+    LOG4CXX_WARN(logger_, "Request module type & data mismatch!");
```
Comment (AKalinich-Luxoft) [inline src/components/can_cooperation/src/commands/button_press_request.cc]: @LitvinenkoIra better to say "Requested module type and data mismatch" here and in `SendResponse`

File: src/components/can_cooperation/src/commands/button_press_request.cc [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                                  (button_name == enums_value::kRecirculate) ||
+                                  (button_name == enums_value::kFanUp) ||
+                                  (button_name == enums_value::kFanDown) ||
+                                  (button_name == enums_value::kTempUp) ||
+                                  (button_name == enums_value::kTempDown) ||
+                                  (button_name == enums_value::kDefrostMax) ||
+                                  (button_name == enums_value::kDefrost) ||
+                                  (button_name == enums_value::kDefrostRear) ||
+                                  (button_name == enums_value::kUpperVent) ||
+                                  (button_name == enums_value::kLowerVent);
```
Comment (AKalinich-Luxoft) [inline src/components/can_cooperation/src/commands/button_press_request.cc]: @LitvinenkoIra propose to move this comparison snake to a separate class function and then just use
`const bool module_type_and_button_name = SomeFunctionToCheckButtonName();`

File: src/components/can_cooperation/include/can_cooperation/commands/button_press_request.h [hunk 3]
```diff
@@ -70,13 +71,8 @@ class ButtonPressRequest : public BaseCommandRequest {
...
   virtual ~ButtonPressRequest();
 
-  /**
-   * @brief executes specific message validation
-   */
-  virtual bool Validate();
-
  protected:
-  virtual std::string ModuleType(const Json::Value& message);
+  virtual std::string ModuleType(const Json::Value& message) FINAL;
```
Comment (AKalinich-Luxoft) [inline src/components/can_cooperation/include/can_cooperation/commands/button_press_request.h]: @LitvinenkoIra `virtual` is redundant here

File: src/components/interfaces/HMI_API.xml [hunk 4]
```diff
@@ -4731,12 +4731,42 @@
     </param>
 </function>
 
-
 <function name="OnInteriorVehicleData" functionID="OnInteriorVehicleDataID" messagetype="notification">
   <param name="moduleData" type="Common.ModuleData">
   </param>
 </function>
 
+<function name="ButtonPress" functionID="ButtonPressID" messagetype="request">
```
Comment (AKalinich-Luxoft) [inline src/components/interfaces/HMI_API.xml]: @LitvinenkoIra please compare `ButtonPress` function you are adding with it latest version in [mobile](https://github.com/yang1070/sdl_core/blob/master/src/components/interfaces/MOBILE_API.xml) and [hmi](https://github.com/yang1070/sdl_core/blob/master/src/components/interfaces/HMI_API.xml) apis.
For example here in request should be also `appID` parameter and response should be without params at all...

**Review by AKalinich-Luxoft — CHANGES_REQUESTED** [review]

File: src/components/can_cooperation/include/can_cooperation/commands/button_press_request.h (hunk 3 above)
Comment (LuxoftAKutsan) [inline src/components/can_cooperation/include/can_cooperation/commands/button_press_request.h]: @LitvinenkoIra remove virtual

File: src/components/can_cooperation/src/commands/button_press_request.cc [hunk 5]
```diff
@@ -0,0 +1,129 @@
+/*
+ Copyright (c) 2013, Ford Motor Company
```
Comment (LuxoftAKutsan) [inline src/components/can_cooperation/src/commands/button_press_request.cc]: @LitvinenkoIra 2017

File: src/components/can_cooperation/src/commands/button_press_request.cc [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+void ButtonPressRequest::Execute() {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  const Json::Value request_params =
+      MessageHelper::StringToValue(message_->json_message());
+
+  const std::string button_name = request_params[kButtonName].asString();
+  const std::string module_type = request_params[kModuleType].asString();
+  bool module_type_and_button_name = true;
```
Comment (LuxoftAKutsan) [inline src/components/can_cooperation/src/commands/button_press_request.cc]: @LitvinenkoIra button_name_matches_module_type

File: src/components/interfaces/MOBILE_API.xml [hunk 7]
```diff
@@ -2614,9 +2614,11 @@
 <!--
   Reverse SDL specific ids, hardcoded in functional module
 -->
-    <element name="GetInteriorVehicleDataID" value="100017" hexvalue="186B1" />
-    <element name="SetInteriorVehicleDataID" value="100018" hexvalue="186B2" />
-    <element name="OnInteriorVehicleDataID" value="100019" hexvalue="186B3" />
+    <element name="ButtonPressID" value="41" hexvalue="29" />
+    <element name="GetInteriorVehicleDataID" value="43" hexvalue="2B" />
+    <element name="SetInteriorVehicleDataID" value="44" hexvalue="2C" />
+
```
Comment (LuxoftAKutsan) [inline src/components/interfaces/MOBILE_API.xml]: @LitvinenkoIra empty line redundant

File: src/components/interfaces/HMI_API.xml [hunk 8]
```diff
@@ -4731,12 +4731,42 @@
...
+  </param>
+  <param name="buttonName" type="Common.ButtonName">
+    <description>The name of supportted RC climate or radio button.</description>
+  </param>
+  <param name="buttonPressMode" type="Common.ButtonPressMode">
+    <description>Indicates whether this is a LONG or SHORT button press event.</description>
+  </param>
+</function>
+
+<function name="ButtonPress" functionID="ButtonPressID" messagetype="response">
```
Comment (LuxoftAKutsan) [inline src/components/interfaces/HMI_API.xml]: @LitvinenkoIra add @OlesiaV  to review

**Review by LuxoftAKutsan — CHANGES_REQUESTED** [review]

File: src/components/can_cooperation/src/commands/button_press_request.cc [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+void ButtonPressRequest::OnEvent(
+    const can_event_engine::Event<application_manager::MessagePtr, std::string>&
+        event) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  DCHECK_OR_RETURN_VOID(
+      (functional_modules::hmi_api::button_press == event.id()));
+
+  application_manager::Message& hmi_response = *(event.event_message());
+  const bool validate_result = application_manager::SUCCESS ==
```
Comment (AByzhynar) [inline src/components/can_cooperation/src/commands/button_press_request.cc]: @LitvinenkoIra Can not see further usage of this variable...

**Review by AByzhynar — CHANGES_REQUESTED** [review]

**Review by OlesiaV — APPROVED** [review]

File: src/components/can_cooperation/src/commands/button_press_request.cc (hunk 2 above)
Comment (LitvinenkoIra) [reply src/components/can_cooperation/src/commands/button_press_request.cc]: @AKalinich-Luxoft Done
9402dc90b74d4c3d4932b95768ab1514f813f50d

File: src/components/can_cooperation/src/commands/button_press_request.cc (hunk 1 above)
Comment (LitvinenkoIra) [reply src/components/can_cooperation/src/commands/button_press_request.cc]: @AKalinich-Luxoft Done
9402dc90b74d4c3d4932b95768ab1514f813f50d

File: src/components/can_cooperation/include/can_cooperation/commands/button_press_request.h (hunk 3 above)
Comment (LitvinenkoIra) [reply src/components/can_cooperation/include/can_cooperation/commands/button_press_request.h]: @AKalinich-Luxoft Done
9402dc90b74d4c3d4932b95768ab1514f813f50d

File: src/components/interfaces/HMI_API.xml (hunk 4 above)
Comment (LitvinenkoIra) [reply src/components/interfaces/HMI_API.xml]: @AKalinich-Luxoft `ButtonPress` already present in HMI_API and MOBILE_API updated according to the latest version
9402dc90b74d4c3d4932b95768ab1514f813f50d

**Comment by LitvinenkoIra** [comment]
@LuxoftAKutsan @AByzhynar please review the latest changes

File: src/components/can_cooperation/src/commands/button_press_request.cc (hunk 1 above)
Comment (AKalinich-Luxoft) [reply src/components/can_cooperation/src/commands/button_press_request.cc]: @LitvinenkoIra can't see changes regarding this issue

**Review by LuxoftAKutsan — APPROVED** [review]

File: src/components/interfaces/MOBILE_API.xml [hunk 10]
```diff
@@ -2617,6 +2617,7 @@
     <element name="GetInteriorVehicleDataID" value="100017" hexvalue="186B1" />
     <element name="SetInteriorVehicleDataID" value="100018" hexvalue="186B2" />
     <element name="OnInteriorVehicleDataID" value="100019" hexvalue="186B3" />
+    <element name="ButtonPressID" value="100015" hexvalue="186B0" />
```
Comment (AByzhynar) [inline src/components/interfaces/MOBILE_API.xml]: @LitvinenkoIra Looks that hex value must be `186AF`

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: src/components/interfaces/MOBILE_API.xml (hunk 10 above)
Comment (LitvinenkoIra) [reply src/components/interfaces/MOBILE_API.xml]: @AByzhynar Done 5237634901076c51319a3fc8ee5d9741d5a031ab

**Review by AByzhynar — APPROVED** [review]

**Review by AKalinich-Luxoft — APPROVED** [review]

**Pull request merged**